### PR TITLE
Optimizing timeouts and packet fragmentation in SNMP scrape jobs

### PR DESCRIPTION
**Pull request by maunavarro**
@SuperQ @RichiH This PR proposes a solution to address the `snmp_exporter` issue of packets being fragmented in UDP. At higher max repetition values (above 1450 MTU), timeouts become more frequent, and at lower max values they can result in timeouts due to increased polling time. The solution being proposed tunes the `MaxRepetitions` value to eliminate/reduce timeout failures.

### CODE REVIEW

**Review by SuperQ — CHANGES_REQUESTED** [review]
This seems to be adding an infinite loop around the existing retry logic. I'm not sure that's what we want.

This looks to be potentially a very bad load generator in the case where you have a long polling time (1m).

**Comment by SuperQ** [comment]
While I get the desire to automatically handle `MaxRepetitions` with a backoff, this seems like it would be a very bad denial of service generator for a lot of cases.

IMO, we should have metrics and/or logs that report back to the user that they're hitting issues with packet MTU. This would allow them to permanently fix the issue in their configuration, rather than rely on the backoff loop every scrape.

**Comment by maunavarro** [comment]
@SuperQ  Came up with code that fragments max repetitions to x numbers to prevent too many loops that can lead to denial of service attacks.

For example:
- if maxRepetitions = 10 and maxRetries = 10 then fragmentMaxRepetitions will be 9,8,7,6,5,4,3,2,1,0
- if maxRepetitions = 5 and maxRetries = 10 then fragmentMaxRepetitions will be 4,3,2,1,0
- if maxRepetitions = 20 and maxRetries = 10 then fragmentMaxRepetitions will be 18,16,14,12,10,8,6,4,2,0
- if maxRepetitions = 1000 and maxRetries = 10 then fragmentMaxRepetitions will be 900,800,700,600,500,400,300,200,100,0

What's your take on this one?

File: collector/collector.go [hunk 1]
```diff
@@ -83,11 +83,46 @@ type ScrapeResults struct {
 }
 
 func ScrapeTarget(ctx context.Context, target string, auth *config.Auth, module *config.Module, logger log.Logger, metrics Metrics) (ScrapeResults, error) {
+
+	timeSleep := time.Second * 1
```
[the remark this answers is not in the document]
Comment (maunavarro) [reply collector/collector.go]: Removed this.

[the remark this answers is not in the document]
Comment (maunavarro) [reply collector/collector.go]: Revamped this whole section.

File: collector/collector.go [hunk 2]
```diff
@@ -83,11 +83,46 @@ type ScrapeResults struct {
...
+		results, timeout, err := scrapeTarget(ctx, target, auth, module, logger, metrics, maxRepetitions)
+		if err != nil {
+			//retry if error is timeout and module walk_retry_enabled config has been set to true
+			if timeout && module.WalkRetryEnabled {
+
+				//return error if last maxRepetitions equals to 0
+				if maxRepetitions == 0 {
+					return results, err
+				}
+				maxRepetitions--
```
[the remark this answers is not in the document]
Comment (maunavarro) [reply collector/collector.go]: Revamped this whole section.

**Comment by SuperQ** [comment]
Sorry, I don't think this is a change we want to accept. I just can't see any up-side to doing this kind of dynamic backoff.

* It's excessive complexity.
* For devices that need backoff, it will always send too-large requests first, leading to excessive load.

You would be better off classifying the capabilities of your devices and pre-configuring a valid max repetitions.

**Comment by SuperQ** [comment]
Ping @roidelapluie or @RichiH, any opinions on this?

**Comment by RichiH** [comment]
I am very much torn. I can see the appeal to solve a specific problem, but it's also violating the principle of least surprise.

At the same time, I don't see how snmp_exporter can reliably detect and report back on this unless it starts changing MTU sizes. Some baseline detection and a helper program to facilitate debugging might be an option.

To unblock @maunavarro maybe an experimental feature flag might be viable? Don't love it, but would allow us to gather some sort of real-world feedback.

**Comment by SuperQ** [comment]
Sorry, I am going to have reject this change.

Because there is no good feedback loop that avoids always starting with incorrect max repetitions, this is always going to cause excessive traffic on devices where it's a problem. There's also no good way to provide feedback to the user that this is a problem because timeouts happen for a variety of reasons (network problems, auth problems, etc).

This change is just going to result in unexpected amounts of excess traffic, and will make it harder to debug.

**Pull request closed without merging**